### PR TITLE
Episode VI: Return of the Promise PRs (getPage Promise support)

### LINKED ISSUE #425: getNextPage not correctly promisified
When using getNextPage, the function is returning before the http request is finished

### PR DESCRIPTION
**Pull request by jamsinclair**
Was hesitant to re-open the issue you just closed #425. I had a similar situation where I was trying to retrieve a large number or all records via paging and promises.  

This PR follows in the footsteps of #262 and adds pluggable promise support to all `getPage` based methods.

Not as nice as I would've liked. Most of it is duplicated code taken from the #262 solution.
## Example outcome from PR:

``` javascript
var GitHubApi = require('github');

var gh = new GitHubApi ({
  Promise: require('bluebird')
});

function getAllOrgRepos(orgName) {
  var repos = [];

  function pager(res) {
    repos = repos.concat(res);
    if (gh.hasNextPage(res)) {
      return gh.getNextPage(res)
        .then(pager);
    }
    return repos;
  }

  return gh.repos.getForOrg({ org: orgName })
    .then(pager);
}


getAllOrgRepos('organization')
  .then(/* Do stuff with all repos */);
```

If this is not at all how you would like it approached feel free to close this PR, as not to waste your time 😸. Very happy to make changes.

Edit: Tidied up example code.

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -414,9 +414,12 @@ var Client = module.exports = function(config) {
     };
 
     function getPage(link, which, headers, callback) {
+        var self = this;
         var url = getPageLinks(link)[which];
-        if (!url)
-            return callback(new error.NotFound("No " + which + " page found"));
+        if (!url) {
+            var urlErr = new error.NotFound("No " + which + " page found");
+            return self.Promise && !callback ? self.Promise.reject(urlErr) : callback(urlErr);
```
Comment (jamsinclair) [inline lib/index.js:421]: Worth throwing an error if neither promise or callback defined, like below?

Comment (kaizensoze) [reply lib/index.js:421]: Up to you. Not a dealbreaker without it.

**Comment by kaizensoze** [comment]
This looks good. I'll add the example output to the examples section. As a bonus, how much work would it be to get the promise stuff also compatible with Q? If a significant amount I'll just go ahead and merge this.

**Comment by jamsinclair** [comment]
From a quick 5 minute play, _it seems_ like the Q library is compatible. I've commented in the issue #400 😄 

If you're fine with the PR, happy for you to merge 👍

File: lib/index.js (hunk 1 above)
Comment (jamsinclair) [reply lib/index.js:421]: I'll leave it be 😄 , I'm sure my lazy developer habits will come back to haunt me.

**Comment by kaizensoze** [comment]
Awesome. Good to go.

**Pull request merged**